### PR TITLE
[BREAK] Remove cache layer and internal calculated property `room.usernames`

### PR DESCRIPTION
**Pull request by rodrigok**
This is a test case, could be or could not be merged depending on tests results.

Since we introduced our cache layer we have been testing how to improve our performance in edge scenarios and one of the most important part that could be making our performance bad could be the cache layer, that's why we have this PR removing it to be able to run custom builds in certain deployments and track the performance.

## Why we added the cache layer
Some time ago we did some small tests and figure out that the most of our queries would run fast or equal if we do them in memory, that would reduce the stress over the database and would allow us to depend less on oplog since we could reproduce the changes in memory and send the events to the clients without wait by the database oplog or even require the oplog been enabled.

Another positive part was the ability to remove the `room.usernames` from database, which was na array that at that time was of around 150k strings in our room `general` at our open/demo server. Using the cache we was able to remount that array in memory only and keep using it wherever we need.

Yet another good part was the ability to have relations in memory where we could find all room of a single user without need to send a query containing an array of thousands of ids to the database since we need to find all the subscriptions, extract the `rid` properties and then find by them.

We had some good results using that approach.

## Why we decide to test the cache removal
Every time our instances crash we need to wait several minutes to them be restarted since the cache load will move 200k users records to memory, 4.7mi of subscriptions and 130k rooms, that stops the process by more then 5 minutes, create a huge load in database and increase the process memory to more than 2MB at the beginning. We force the CPU to deal with that and need to run the relationship methods over that amount of data, all that was causing a problem when we have outages since we could not bring all cloud users back at the same time or our cache load would crash our database.

We found some evidences that we need more instances since each one is busy doing queries in memory or even rebuild the relations when data changes, when we have a few users using the system the cache would be much better then the database in response time, but when we start having more users the CPU of the process start to be bottleneck while the database is free of usage.

We removed the `room.usernames` at all, since that amount of data in an array does not make sense, of course that forced us to back the big queries and some denormalization, our main goal here is to test that on our open server and see how the processes and database will deal with that change. We will introduce some other important performance improvements later to reduce the amount of data we process for notifications.

## How it will affect the forks
Well, we will have a document listing all the APIs we removed and we will try to keep the changes compatible, one of the things we will not be able to keep will be the `room.usernames` array, if you rely on that we could discuss how to deal with the change.

## Work in progress
As soon we have more information and metrics we will update this description.

## Breaking Changes
- Method `public-settings/get` doesn't return the field `type` anymore.

### CODE REVIEW

**Comment by graywolf336** [comment]
@rodrigok I know this is a work in progress, but can we get some description about what this is, why it's being done, and how it effects those who have custom forks?

**Comment by rodrigok** [comment]
@graywolf336 done, thanks by the reminder, I was planning to do that

File: server/startup/migrations/v125.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			fields: {
+				_id: 1
+			}
+		}).forEach(({_id}) => {
+			const usersCount = RocketChat.models.Subscriptions.findByRoomId(_id).count();
+
+			RocketChat.models.Rooms.update({
+				_id
+			}, {
+				$set: {
```
Comment (rodrigok) [inline server/startup/migrations/v125.js]: Should use the internal collection to prevent mess the channel's list order

File: packages/rocketchat-api/server/v1/channels.js [hunk 2]
```diff
@@ -553,38 +547,32 @@ RocketChat.API.v1.addRoute('channels.members', { authRequired: true }, {
...
+		const members = RocketChat.models.Subscriptions.findByRoomId(findResult._id, {
+			sort: {'u.username':  sort.username != null ? sort.username : 1},
 			skip: offset,
 			limit: count
-		});
-
-		if (shouldBeOrderedDesc) {
-			members = members.reverse();
-		}
+		}).fetch().map(s => s.u && s.u.username);
```
Comment (ggazzo) [inline packages/rocketchat-api/server/v1/channels.js]: the next query will search for all users without usernames

File: packages/rocketchat-api/server/v1/channels.js [hunk 3]
```diff
@@ -593,16 +581,15 @@ RocketChat.API.v1.addRoute('channels.messages', { authRequired: true }, {
...
+			checkedArchived: false
 		});
 		const { offset, count } = this.getPaginationItems();
 		const { sort, fields, query } = this.parseJsonQuery();
 
 		const ourQuery = Object.assign({}, query, { rid: findResult._id });
 
 		//Special check for the permissions
-		if (RocketChat.authz.hasPermission(this.userId, 'view-joined-room') && !findResult.usernames.includes(this.user.username)) {
+		if (RocketChat.authz.hasPermission(this.userId, 'view-joined-room') && !RocketChat.models.Subscriptions.findOneByRoomIdAndUserId(findResult._id, this.userId, {fields: {_id: 1}})) {
```
Comment (ggazzo) [inline packages/rocketchat-api/server/v1/channels.js]: do you think this way will be worst ?
```javascript
		const ids = online.map((result, {_id}) => _id)

		const online = RocketChat.models.Subscriptions.find({ rid: root._id, _id: { $in: ids } }, { fields: { _id: 1, username: 1 } });
		return RocketChat.API.v1.success({ online });
```

File: packages/rocketchat-api/server/v1/groups.js [hunk 4]
```diff
@@ -449,23 +445,31 @@ RocketChat.API.v1.addRoute('groups.listAll', { authRequired: true }, {
...
 
-		const members = RocketChat.models.Rooms.processQueryOptionsOnResult(Array.from(findResult._room.usernames).sort(sortFn), {
+		const members = RocketChat.models.Subscriptions.findByRoomId(findResult.rid, {
+			fields: {u: 1},
+			sort: subscriptionSort,
 			skip: offset,
 			limit: count
-		});
+		}).fetch()
+			.filter(s => s.u && s.u._id && s.u.username)
```
Comment (ggazzo) [inline packages/rocketchat-api/server/v1/groups.js]: we could use `"u":{$exists:1}..."u._id": {$exists:1} ... "u.username": {$exists:1} `  or not?

File: packages/rocketchat-lib/server/functions/removeUserFromRoom.js [hunk 5]
```diff
@@ -3,9 +3,10 @@ RocketChat.removeUserFromRoom = function(rid, user) {
 
 	if (room) {
 		RocketChat.callbacks.run('beforeLeaveRoom', user, room);
-		RocketChat.models.Rooms.removeUsernameById(rid, user.username);
 
-		if (room.usernames.indexOf(user.username) !== -1) {
+		const subscription = RocketChat.models.Subscriptions.findOneByRoomIdAndUserId(rid, user._id, {fields: {_id: 1}});
+
+		if (subscription != null) {
```
Comment (ggazzo) [inline packages/rocketchat-lib/server/functions/removeUserFromRoom.js]:
```

File: packages/rocketchat-lib/server/methods/addUsersToRoom.js [hunk 6]
```diff
@@ -16,8 +16,8 @@ Meteor.methods({
 		// Get user and room details
 		const room = RocketChat.models.Rooms.findOneById(data.rid);
 		const userId = Meteor.userId();
-		const user = Meteor.user();
-		const userInRoom = Array.isArray(room.usernames) && room.usernames.includes(user.username);
+		const subscription = RocketChat.models.Subscriptions.findOneByRoomIdAndUserId(data.rid, userId, {fields: {_id: 1}});
```
Comment (ggazzo) [inline packages/rocketchat-lib/server/methods/addUsersToRoom.js]: should we use `.count()` ?

**Comment by mrsimpson** [comment]
@rodrigok when removing `room.usernames`, this breaks a couple of places where we needed to grab the members of the room. Having a method on room model level which returns the members would be an alternative.

File: packages/rocketchat-api/server/v1/channels.js (hunk 2 above)
Comment (sampaiodiego) [reply packages/rocketchat-api/server/v1/channels.js]: actually the issue was it returning subscriptions with `null` usernames. I've fixed it in 187dbc527aa078a848e60998c2cb6cce3cde5f82

File: server/startup/migrations/v127.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
+import Future from 'fibers/future';
+
+RocketChat.Migrations.add({
+	version: 126,
+	up() {
+		RocketChat.models.Rooms.update(
```
Comment (rodrigok) [inline server/startup/migrations/v127.js]: Need to use db directly to not update the field `_updatedAt`

File: server/startup/migrations/v127.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
+import Future from 'fibers/future';
+
+RocketChat.Migrations.add({
+	version: 126,
```
Comment (sampaiodiego) [inline server/startup/migrations/v127.js]: migration is `126` while the file name is `127`. what is correct?

File: packages/rocketchat-api/server/v1/channels.js (hunk 3 above)
Comment (rodrigok) [reply packages/rocketchat-api/server/v1/channels.js]: @ggazzo did you comment at the right place?

File: packages/rocketchat-api/server/v1/groups.js (hunk 4 above)
Comment (rodrigok) [reply packages/rocketchat-api/server/v1/groups.js]: We could, need to create a new method `findByRoomIdWhenUserUsernameExists` rather than `findByRoomId`

File: packages/rocketchat-lib/server/functions/removeUserFromRoom.js (hunk 5 above)
Comment (rodrigok) [reply packages/rocketchat-lib/server/functions/removeUserFromRoom.js]: ??? Obviously I prefer what I did 😃

**Comment by rodrigok** [comment]
@mrsimpson seems good, we could add that helper within a deprecation warning, since that will not be performatic

File: packages/rocketchat-api/server/v1/channels.js (hunk 3 above)
Comment (ggazzo) [reply packages/rocketchat-api/server/v1/channels.js]: maybe :p

File: packages/rocketchat-api/server/v1/channels.js [hunk 9]
```diff
@@ -643,7 +634,8 @@ RocketChat.API.v1.addRoute('channels.online', { authRequired: true }, {
 
 		const onlineInRoom = [];
 		online.forEach(user => {
-			if (room.usernames.indexOf(user.username) !== -1) {
+			const subscription = RocketChat.models.Subscriptions.findOneByRoomIdAndUserId(root._id, user._id, {fields: {_id: 1}});
```
Comment (ggazzo) [inline packages/rocketchat-api/server/v1/channels.js]: root?

Comment (rodrigok) [reply packages/rocketchat-api/server/v1/channels.js]: Fixed, thanks

File: server/methods/channelsList.js [hunk 10]
```diff
@@ -44,19 +44,22 @@ Meteor.methods({
...
 
-		const roomTypes = [];
+		let channels = [];
 
 		if (channelType !== 'private') {
 			if (RocketChat.authz.hasPermission(Meteor.userId(), 'view-c-room')) {
-				roomTypes.push({
-					type: 'c'
-				});
+				if (filter) {
```
Comment (ggazzo) [inline server/methods/channelsList.js]: I'm not sure if this test is right, I think we inverted because we are using filter when we dont have filter 🤔

File: packages/rocketchat-api/server/v1/groups.js [hunk 11]
```diff
@@ -453,42 +454,36 @@ RocketChat.API.v1.addRoute('groups.listAll', { authRequired: true }, {
...
+			fields: { 'u._id': 1 },
+			sort: { 'u.username': sort.username != null ? sort.username : 1 },
 			skip: offset,
 			limit: count
-		}).fetch()
-			.filter(s => s.u && s.u._id && s.u.username)
-			.map(s => s.u.username);
+		});
+
+		const total = subscriptions.count();
```
Comment (rodrigok) [inline packages/rocketchat-api/server/v1/groups.js]: Total will count the users without username???

Comment (ggazzo) [reply packages/rocketchat-api/server/v1/groups.js]: sure, I think this code has the same behavior
```javascript
{
	total: RocketChat.models.Subscriptions.findByRoomId(findResult.rid).count()	
}
```

Comment (rodrigok) [reply packages/rocketchat-api/server/v1/groups.js]: I don't think so, you are including the count and skip right?

File: packages/rocketchat-api/server/v1/im.js [hunk 12]
```diff
@@ -275,23 +278,25 @@ RocketChat.API.v1.addRoute(['dm.messages.others', 'im.messages.others'], { authR
...
 		const { offset, count } = this.getPaginationItems();
-		const { sort, fields } = this.parseJsonQuery();
+		const { sort = { name: 1 }, fields } = this.parseJsonQuery();
 
-		// TODO: CACHE: Add Bracking notice since we removed the query param
-		const rooms = RocketChat.models.Rooms.findBySubscriptionTypeAndUserId('d', this.userId, {
-			sort: sort ? sort : { name: 1 },
+		// TODO: CACHE: Add Breacking notice since we removed the query param
+
+		const cursor = RocketChat.models.Rooms.findBySubscriptionTypeAndUserId('d', this.userId, {
```
Comment (rodrigok) [inline packages/rocketchat-api/server/v1/im.js]: Are you counting only the returned items?

File: packages/rocketchat-apps/server/bridges/messages.js [hunk 13]
```diff
@@ -64,13 +64,10 @@ export class AppMessageBridge {
...
 			});
 
-			const users = RocketChat.models.Subscriptions.findByRoomId(room._id, {fields: {u: 1}}).fetch().filter(s => s.u && s.u._id).map(s => s.u);
-			users.forEach((u) => {
-				const user = RocketChat.models.Users.findOneById(u._id);
-				if (user) {
-					RocketChat.Notifications.notifyUser(user._id, 'message', rmsg);
-				}
-			});
+			const users = RocketChat.models.Subscriptions.find({ rid: room._id, u: { $exists: 1 }, 'u._id': { $exists: 1 } }, { fields: { 'u._id': 1 } })
```
Comment (rodrigok) [inline packages/rocketchat-apps/server/bridges/messages.js]: Move this to model `findByRoomIdWhenUserIdExists`

File: packages/rocketchat-apps/server/bridges/messages.js [hunk 14]
```diff
@@ -64,13 +64,10 @@ export class AppMessageBridge {
...
+				.fetch()
+				.map(s => s.u._id);
+			RocketChat.models.Users.find({ _id: { $in: users } }).fetch().forEach(({ _id }) => RocketChat.Notifications.notifyUser(_id, 'message', rmsg));
```
Comment (rodrigok) [inline packages/rocketchat-apps/server/bridges/messages.js]: Move this to model `findByIds` and return only the field `_id`

File: packages/rocketchat-channel-settings/server/models/Rooms.js [hunk 15]
```diff
@@ -16,27 +16,27 @@ RocketChat.models.Rooms.setReadOnlyById = function(_id, readOnly) {
...
 		$set: {
-			ro: readOnly
+			ro: readOnly,
+			muted: []
 		}
 	};
 	if (readOnly) {
-		RocketChat.models.Subscriptions.findByRoomId(_id).forEach(function(subscription) {
-			if (subscription.u == null || subscription.u._id == null || subscription.u.username == null) {
+		RocketChat.models.Subscriptions.find({ rid: _id, 'u._id' : { $exists : 1 }, 'u.username' : { $exists : 1 } }).forEach(function(subscription) {
```
Comment (rodrigok) [inline packages/rocketchat-channel-settings/server/models/Rooms.js]: Move this find to the model

File: packages/rocketchat-graphql/server/resolvers/channels/Channel-type.js [hunk 16]
```diff
@@ -14,9 +14,8 @@ const resolver = {
 			return root.name;
 		},
 		members: (root) => {
-			return RocketChat.models.Subscriptions.findByRoomId(root._id, {fields: {u: 1}}).fetch().filter(s => s.u && s.u._id).map(
-				s => RocketChat.models.Users.findOneById(s.u._id)
-			);
+			const ids = RocketChat.models.Subscriptions.find({ rid: root._id, 'u._id': { $exists: 1 } }, { fields: { u: 1 } }).fetch();
```
Comment (rodrigok) [inline packages/rocketchat-graphql/server/resolvers/channels/Channel-type.js]: Moe these finds to the models

File: packages/rocketchat-lib/server/functions/removeUserFromRoom.js [hunk 17]
```diff
@@ -18,6 +18,7 @@ RocketChat.removeUserFromRoom = function(rid, user) {
 		RocketChat.models.Subscriptions.removeByRoomIdAndUserId(rid, user._id);
 
 		Meteor.defer(function() {
+			// TODO: CACHE: maybe a queue?
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/functions/removeUserFromRoom.js]: What is this?

File: packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js [hunk 18]
```diff
@@ -237,17 +237,15 @@ function sendAllNotifications(message, room) {
...
-				await callJoinRoom(user, room._id);
-
-				return user._id;
-			})
-		).then((users) => {
-			users.forEach((userId) => {
-				const subscription = RocketChat.models.Subscriptions.findOneByRoomIdAndUserId(room._id, userId);
-
+		const mentions = message.mentions.filter(({ _id }) => _id !== 'here' && _id !== 'all').map(({ _id }) => _id);
+		Promise.all(RocketChat.models.Subscriptions.find({
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js]: Move this find to the models

File: packages/rocketchat-lib/server/models/Rooms.js [hunk 19]
```diff
@@ -9,11 +9,14 @@ class Document {
 	get usernames() {
 		// TODO: Remove deprecation on version 0.70
 		console.warn('DEPRECATION: Room.usernames is deprecated and will be removed on version 0.70.0');
-		return RocketChat.models.Subscriptions.findByRoomId(this._id, {
+		return RocketChat.models.Subscriptions.find({
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/models/Rooms.js]: Move this find to the model

File: packages/rocketchat-slashcommands-inviteall/server/server.js [hunk 20]
```diff
@@ -32,19 +32,19 @@ function inviteAll(type) {
 				}, currentUser.language)
 			});
 		}
-
-		const users = RocketChat.models.Subscriptions.findByRoomId(baseChannel._id, {fields: {u: 1}}).fetch().map(s => s.u && s.u.username);
+		const cursor = RocketChat.models.Subscriptions.find({ rid: baseChannel._id, 'u.username': { $exists: 1 } }, { fields: { 'u.username': 1 } });
```
Comment (rodrigok) [inline packages/rocketchat-slashcommands-inviteall/server/server.js]: Move this find to the model

File: server/methods/channelsList.js [hunk 21]
```diff
@@ -63,20 +65,19 @@ Meteor.methods({
...
 				fields: {
 					username: 1,
 					'settings.preferences.sidebarGroupByType': 1
 				}
 			});
 			const userPref = RocketChat.getUserPreference(user, 'sidebarGroupByType');
-			const globalPref = RocketChat.settings.get('UI_Group_Channels_By_Type');
 			// needs to negate globalPref because userPref represents its opposite
-			const groupByType = userPref !== undefined ? userPref : globalPref;
+			const groupByType = userPref !== undefined ? userPref : RocketChat.settings.get('UI_Group_Channels_By_Type');
```
Comment (rodrigok) [inline server/methods/channelsList.js]: Is this change related to cache?

File: server/methods/getUsersOfRoom.js [hunk 22]
```diff
@@ -1,21 +1,22 @@
...
 
-		if (room.broadcast && !RocketChat.authz.hasPermission(Meteor.userId(), 'view-broadcast-member-list', roomId)) {
+		if (room.broadcast && !RocketChat.authz.hasPermission(userId, 'view-broadcast-member-list', rid)) {
 			throw new Meteor.Error('error-not-allowed', 'Not allowed', { method: 'getUsersOfRoom' });
 		}
 
-		const subscriptions = RocketChat.models.Subscriptions.findByRoomId(roomId, {fields: {u: 1}}).fetch().filter(s => s.u && s.u._id && s.u.username);
-		const userIds = subscriptions.map(s => s.u._id);
-		const options = {fields: {username: 1, name: 1}};
+		const subscriptions = RocketChat.models.Subscriptions.find({ rid, 'u.username': { $exists: 1 } }, { fields: { 'u._id': 1 } }).fetch();
```
Comment (rodrigok) [inline server/methods/getUsersOfRoom.js]: Move this find to the model

File: packages/rocketchat-api/server/v1/groups.js (hunk 11 above)
Comment (ggazzo) [reply packages/rocketchat-api/server/v1/groups.js]: `count()` ignores `limit` and `skip`
